File: packages/emitter/testcases/common/functions/async-hof/AsyncReturningFunctions.ts
import { int } from "@tsonic/core/types.js";

// Async function returning sync function
export async function createMultiplier(factor: int): Promise<(x: int) => int> {
  return x => x * factor;
}

// Async function returning async function
export async function createAsyncAdder(
  base: int
): Promise<(x: int) => Promise<int>> {
  return async x => base + x;
}

// Function taking async callback
export async function withAsyncCallback<T>(
  value: T,
  callback: (x: T) => Promise<string>
): Promise<string> {
  return await callback(value);
}

File: packages/emitter/testcases/common/expected/functions/async-hof/AsyncReturningFunctions.cs
namespace TestCases.common.functions.asynchof
{
    [global::Tsonic.Internal.ModuleContainerAttribute]
    public static class AsyncReturningFunctions
    {
        public static async global::System.Threading.Tasks.Task<global::System.Func<int, int>> createMultiplier(int factor)
        {
            return (int x) => x * factor;
        }

        public static async global::System.Threading.Tasks.Task<global::System.Func<int, global::System.Threading.Tasks.Task<int>>> createAsyncAdder(int @base)
        {
            return async (int x) => @base + x;
        }

        public static async global::System.Threading.Tasks.Task<string> withAsyncCallback<T>(T value, global::System.Func<T, global::System.Threading.Tasks.Task<string>> callback)
        {
            return await callback(value);
        }
    }
}